icu4c/source/test/intltest/messageformat2test.cpp

```diff
@@ -263,6 +265,36 @@ void TestMessageFormat2::testAPICustomFunctions() {
     delete person;
 }
 
+// ICU-22890 lone surrogate cause infinity loop
+void TestMessageFormat2::testHighLoneSurrogate() {
+    IcuTestErrorCode errorCode(*this, "testHighLoneSurrogate");
+    UParseError pe = { 0, 0, {0}, {0} };
+    // Lone surrogate with only high surrogate
+    UnicodeString loneSurrogate({0xda02, 0});
+    icu::message2::MessageFormatter msgfmt1 =
+      icu::message2::MessageFormatter::Builder(errorCode)
+      .setPattern(loneSurrogate, pe, errorCode)
+      .build(errorCode);
+    UnicodeString result = msgfmt1.formatToString({}, errorCode);
+    assertEquals("testHighLoneSurrogate", U_MF_SYNTAX_ERROR, errorCode);
+    errorCode.reset();
```

Comment:
why do we reset the errorCode here?

Reply:
I thought that in intltest, each test has to reset the error code so it doesn't carry over to the next test -- is that not the case?

Reply:
no, errorCode is a local variable in this test.

Reply:
please remove the reset

Reply:
If I remove the `errorCode.reset()` call, then the test fails. I assume this has to do with the reference to `this`:

```
    IcuTestErrorCode errorCode(*this, "testHighLoneSurrogate");
```

and that there's some shared state that results in an error if the error code is non-success at the end of a test.

A lot of other tests in intltest have this pattern (`errorCode` declared as a local `IcuTestErrorCode` variable, and then `errorCode.reset()` at the end of the method.)

Reply:
so sorry, I got confused, it should be 
```
errorCode.expectErrorAndReset(U_MF_SYNTAX_ERROR, "testHighLoneSurrogate");
```
instead of
```
assertEquals("testHighLoneSurrogate", U_MF_SYNTAX_ERROR, errorCode);
errorCode.reset();
```
same below

Reply:
OK, fixed in [a9a6de2](https://github.com/unicode-org/icu/pull/3167/commits/a9a6de272b531fd63818f2334d4f58f244589103)